internal/pkg/vm/internal/vmcontext/vmcontex2.go

```diff
@@ -65,15 +66,80 @@ type actorStorage struct {
 }
 
 // NewVM creates a new runtime for executing messages.
-func NewVM(rnd RandomnessSource, actorImpls ActorImplLookup, store storage.VMStorage, st state.Tree) VM {
+func NewVM(rnd RandomnessSource, actorImpls ActorImplLookup, store *storage.VMStorage, st state.Tree) VM {
 	return VM{
-		rnd:        rnd,
-		actorImpls: actorImpls,
-		store:      store,
-		state:      state.NewCachedTree(st),
+		rnd:          rnd,
+		actorImpls:   actorImpls,
+		store:        store,
+		state:        state.NewCachedTree(st),
+		currentEpoch: *types.NewBlockHeight(0),
 	}
 }
 
+// ApplyGenesisMessage forces the execution of a message in the vm actor.
+//
+// This method is intended to be used in the generation of the genesis block only.
+func (vm *VM) ApplyGenesisMessage(from address.Address, to address.Address, method types.MethodID, value types.AttoFIL, params ...interface{}) (interface{}, error) {
+	// get the params into bytes
+	encodedParams, err := abi.ToEncodedValues(params...)
+	if err != nil {
+		return nil, err
+	}
+
+	// normalize from addr
+	from = vm.normalizeFrom(from)
+
+	// build internal message
+	imsg := internalMessage{
+		miner:  address.Undef,
+		from:   from,
+		to:     to,
+		value:  value,
+		method: method,
+		params: encodedParams,
+	}
+
+	ret, err := vm.applyImplicitMessage(imsg)
+	if err != nil {
+		return ret, err
+	}
+
+	// commit state
+	// flush all objects out
+	if err := vm.store.Flush(); err != nil {
+		return nil, err
+	}
+	// commit new actor state
+	if err := vm.state.Commit(context.Background()); err != nil {
+		return nil, err
+	}
+	// TODO: update state root (issue: #3718)
+
+	return ret, nil
+}
+
+func (vm *VM) normalizeFrom(from address.Address) address.Address {
```

Comment:
Question: do we plan to move to this from the init actor lookup stuff that used to live in the processor across the board?

Reply:
No, this is internal to the VM. 
The reason some places were doing the processor->init-lookup call to go between ID's was to load/set the state for an actor on the outside of the VM. 

The VM layer should provide a wrapper for the tree that lets you grab a view of the state of any actor. That wrapper can can figure out based on the address if it needs to normalize it or not. But that code lives inside the vm and nothing outside needs to worry about what address types we might be using to index the actors.